src/python/pants/rules/core/fmt.py

```diff
@@ -118,40 +148,64 @@ class Fmt(Goal):
 @goal_rule
 async def fmt(
     console: Console,
-    targets_with_origins: HydratedTargetsWithOrigins,
+    targets_with_origins: TargetsWithOrigins,
     options: FmtOptions,
     workspace: Workspace,
     union_membership: UnionMembership,
 ) -> Fmt:
-    adaptors_with_origins = [
-        TargetAdaptorWithOrigin.create(target_with_origin.target.adaptor, target_with_origin.origin)
-        for target_with_origin in targets_with_origins
-        if target_with_origin.target.adaptor.has_sources()
-    ]
-
-    all_language_formatters: Iterable[Type[LanguageFormatters]] = union_membership.union_rules[
-        LanguageFormatters
-    ]
+    language_target_collection_types: Iterable[Type[LanguageFmtTargets]] = (
+        union_membership.union_rules[LanguageFmtTargets]
+    )
+
+    language_target_collections = tuple(
+        language_target_collection_type(
+            TargetsWithOrigins(
+                target_with_origin
+                for target_with_origin in targets_with_origins
+                if language_target_collection_type.belongs_to_language(target_with_origin.target)
+            )
+        )
+        for language_target_collection_type in language_target_collection_types
+    )
+    targets_with_sources = await MultiGet(
+        Get[TargetsWithSources](
+            TargetsWithSourcesRequest(
+                target_with_origin.target
+                for target_with_origin in language_target_collection.targets_with_origins
+            )
+        )
+        for language_target_collection in language_target_collections
+    )
+    # NB: We must convert back the generic TargetsWithSources objects back into their
+    # corresponding LanguageFmtTargets, e.g. back to PythonFmtTargets, in order for the union
+    # rule to work.
+    valid_language_target_collections = tuple(
+        language_target_collection_cls(
+            TargetsWithOrigins(
+                target_with_origin
+                for target_with_origin in language_target_collection.targets_with_origins
+                if target_with_origin.target in language_targets_with_sources
+            )
+        )
+        for language_target_collection_cls, language_target_collection, language_targets_with_sources in zip(
+            language_target_collection_types, language_target_collections, targets_with_sources
+        )
+        if language_targets_with_sources
```

Comment:
This is a little weird because we need to go back from `Targets` to `TargetsWithOrigins`, then to go from `TargetsWithOrigins` to something like `PythonFmtTargets`.

The call site could be a little simpler if we had a way to `await Get[TargetsWithOriginsWithSources](TargetsWithOriginsWithSourcesRequest)`, so that we don't need to go back from `Targets -> TargetsWithOrigins`. But, that is obviously horrible, so I chose the lesser of two evils.

Reply:
I might suggest adding explicit type annotations on these variables, to make it easier for people maintaining this in the future to reason about what these nested tuples are.

Reply:
Good idea.